### DISCUSSION TITLE
Add argument to MBSworker type to allow passing in pointer to memory space

### DISCUSSION THREAD

**Discussion by BrotherCreamy** [eModbus/eModbus#258] [General]
I might be mistaken but at the moment it seems like the only way to practically use the library is to create your own FCxx worker functions which internally reference an external memory space, as in the example code.

I think it makes more sense to pass in a pointer to that memory space, which would allow us to make a higher level library with all of the function codes already defined.

And so the higher level interface would be really straightforward - just create a server, specifying memory space (and address) and start it - e.g.:

```
ModbusServerRTU MyServer(Serial0, TIMEOUT, RTS_PIN, &memSpace, ADDRESS);
MyServer.start();
```

...and that would deploy and start the worker functions for all the function codes.
 
 Is there any reason why this can't work?

Cheers,
BrotherCreamy

**Comment by Miq1** [comment]
I am wondering what would be the advantage of your proposal. If you will pass the ``memspace`` pointer with the server initialization, it would be nothing else but another way to access a fixed external memory area - for the cost of additional methods to fail if the pointer may get invalid at some point. Using global variables will at least save you from that problem.

Get me right: I may just still be misunderstanding what you are planning to achieve. If so, please try to further explain 😉

**Reply by BrotherCreamy** [reply]
My ultimate goal is to make the library simpler to use.

Maybe I'm missing something, but I don't see the point of making the user implement their own worker functions every time.

To my mind, the purpose of Modbus RTU server side is to implement remotely accessible register space. So the user need only define the hardware serial and memory area to be used, and the library should automatically handle the rest.

I'm not sure what the worker functions achieve by having two layers of abstraction on the function codes, e.g.:

```
registerWorker(MY_SERVER, READ_HOLD_REGISTER, &FC03);
```

I guess I'm missing something here.

Anyway let me know what you think.
I will probably eventually adapt the library for my own use, as I like to use a lot of Modbus (old habits die hard), but I will probably butcher it in the process. However, if you think it's a good idea to build a comprehensive Modbus server stack then I am happy to try and help out.

Cheers,
BrotherCreamy

**Comment by Miq1** [comment]
I see. Of course feel free to modify whatever you will find suitable - I surely do not object in any way.

The library is meant to be as flexible as possible to fit to every imaginable application. Yours - a set memory area to be delivered on request - is just one of these. I f.i. have some servers running that will gather data from sensors attached upon a request only; there is no such thing as a memory area to hold the data intermediately. 

How about you writing an add-on library that just wraps the necessary eModbus mechanics with something simpler that does as you are intending?
